### ISSUE TITLE
Backup/restore to specific package ID of an app

### ISSUE THREAD

**Issue by ocdtrekkie** [sandstorm-io/sandstorm#386]
This is a wishlist item of mine I kinda feel merits 'issue' status. Namely because I am prone to install an app, upgrade my files, and then discover whether or not the app update works correctly with those files or if I like the update. (And as more app authors get involved in Sandstorm, the likelihood someone will create a breaking app update at some point will increase.) This suggested flow would allow people to backup their files, upgrade apps and files, and restore original versions of those files if they wish.

The notion I see of how this would work is as follows:

The metadata in a grain backup should not only contain the appId, but contain the packageId as well. (appVersion could do as well, though as packageId is somewhat more absolute to ensuring the restore is exact it's preferable if possible.)

When restoring a file, Sandstorm should compare the grain backup's packageId against the version of the app "installed" on the user's Sandstorm instance. If it is a match, it should restore the file without further question.

However, if the packageId is not a match to the user's installed version, Sandstorm should ask the user if they would like to restore the grain with their currently installed version, or the version that was originally associated with the grain. If the user is fine with using their currently installed version, do so. If the user wants the grain's original version of the app, see if the Sandstorm server has the correct packageId for the app, and if so, use it. (Otherwise Sandstorm should probably query either an app store or repository to see if that version of the package is available, though that extends beyond the scope of this request.)

**Comment by ocdtrekkie** [comment]
As a clarification, appVersion is already stored in metadata (though not readable when I open it in Notepad, unlike the appId).

Current functionality is to disable restore if the grain is newer than the installed version. And if the grain is older or the same, to simply use the installed version. I would prefer if the installed version and grain version do not match, to offer to either use the installed version, or look for the applicable version on the server, regardless of whether the version the user has currently set as "installed" is newer or older.

**Comment by kentonv** [comment]
I can see the value of this, but I think we need to find a different way to do it than a prompt during upload. The problem with such a prompt is that 99% of the time users will want to use the latest version of the app. Since backups are often from the past, it will be pretty common for the backup to have been created using an older version, forcing the prompt. Requiring the user to answer a question when we know what the answer will be 99% of the time is not great UX. We need to have some separate setting that power users can seek out when they want it.

So let me ask this: If the problem is that you may be unhappy with changes in a newer version of the app, then wouldn't you actually want to pin the app to an older version globally? That is to say, when would you ever want backups to be restored with a different version of the app than the one that you get when you click "new document" (assuming you can control the latter, which right now is admittedly non-trivial)?

If these should match, then the answer seems straightforward: We should implement a "pin version" feature that lets you specify your preferred version, and use that both for "new" and for uploaded backups.

Thoughts?

**Comment by ocdtrekkie** [comment]
I am okay with a pinned version concept. I suppose if you need to do a one time oddity, you could change your pinned version, restore, and change back. 

I think the dialog may still be the way to go for restores newer than your installed/pinned version though. As using your pinned version on a newer file may be breaking, but you don't want to force an update on someone's whole account for them to read that one file.

**Comment by ocdtrekkie** [comment]
Other possible note: How is a user to know which version their backup originally used with the pinned method? As it may be critical or desired to ensure the exact experience is restored.

**Comment by kentonv** [comment]
Yes, I agree we should not downgrade an app if at all possible.

> How is a user to know which version their backup originally used with the pinned method? As it may be critical or desired to ensure the exact experience is restored.

Can you give an example where you imagine having the _exact_ version would be critical? My inclination is not to try to handle this unless we see it being a problem for real users in practice.

**Comment by kentonv** [comment]
> Can you give an example where you imagine having the exact version would be critical?

An answer to my own question: When troubleshooting someone else's problems, you might ask them to send you a backup of the grain, and then you probably want to open it using exactly the version they're using. Could be important for app developers in particular.

**Comment by ocdtrekkie** [comment]
I think for the most part being able to select a pinned version and then have all grains, older or newer, try to restore to that version would cover most needs. I think most granular apps have fairly consistent formats from version to version. Draw.io and EtherCalc grains of any age will work with any version of their respective apps for instance. 

I kinda think being able to set specific app versions in specific grains manually may be something that could be done from right click or a menu in the shell perhaps once options like delete and rename are also available in the shell. Just to change the appVersion associated with a given grain. 

That sort of functionality, like the debug log, might be worth putting under some sort of dev/advanced mode though eventually.

**Comment by kentonv** [comment]
Yeah that sounds about right -- except that I'd be very hesitant to downgrade grains without all kinds of red flags and warnings, probably only accessible if you've enabled some "advanced options" thing, since it's pretty likely to break.

**Comment by ocdtrekkie** [comment]
That would be a good place for that prompt. If you have a pinned version older than the grain, rather than failing like it does now, offer to either restore for the pinned version (with warnings) or retrieve newer app version for that grain.

**Comment by zeigerpuppy** [comment]
As a practical example, I am trying to restore a Ghost Blog grain from a previous version and the upload fails with `App id for uploaded grain not installed`.  Is there a simple way to recover the grain in the new install, I presume there's some way to manually install an older version of the app and then upgrade in situ?
